sqlalchemy-stubs/orm/__init__.pyi

```diff
@@ -73,27 +82,148 @@ from .util import polymorphic_union as polymorphic_union
+    comparator_factory: Optional[Any] = ...,
+    single_parent: bool = ...,
+    innerjoin: Union[bool, str] = ...,
+    distinct_target_key: Optional[bool] = ...,
+    doc: Optional[str] = ...,
+    active_history: bool = ...,
+    cascade_backrefs: bool = ...,
+    load_on_pending: bool = ...,
+    bake_queries: bool = ...,
+    _local_remote_pairs: Optional[Any] = ...,
+    query_class: Optional[Any] = ...,
+    info: Optional[MutableMapping[Any, Any]] = ...,
+    omit_join: Optional[Literal[False]] = ...,
+    sync_backref: Optional[Any] = ...,
+) -> Any: ...
+def relation(*arg: Any, **kw: Any) -> Any: ...
+def dynamic_loader(argument: Any, **kw: Any) -> Any: ...
+def column_property(*arg: Any, **kw: Any) -> Any: ...
+def composite(*arg: Any, **kw: Any) -> Any: ...
+@overload
+def mapped_column(
+    __name: str,
+    *args: SchemaEventTarget,
+    autoincrement: Union[bool, Literal["auto", "ignore_fk"]] = ...,
+    default: Optional[Any] = ...,
+    doc: Optional[str] = ...,
+    key: Optional[str] = ...,
+    index: Optional[bool] = ...,
+    info: MutableMapping[Any, Any] = ...,
+    nullable: bool = ...,
+    onupdate: Optional[Any] = ...,
+    primary_key: bool = ...,
+    server_default: Optional[_ServerDefaultType[Any]] = ...,
+    server_onupdate: Optional[FetchedValue] = ...,
+    quote: Optional[bool] = ...,
+    unique: Optional[bool] = ...,
+    system: bool = ...,
+    comment: Optional[str] = ...,
+    **kwargs: Any,
+) -> Any: ...
+@overload
+def mapped_column(
+    *args: SchemaEventTarget,
+    autoincrement: Union[bool, Literal["auto", "ignore_fk"]] = ...,
+    default: Optional[Any] = ...,
+    doc: Optional[str] = ...,
+    key: Optional[str] = ...,
+    index: Optional[bool] = ...,
+    info: MutableMapping[Any, Any] = ...,
+    nullable: bool = ...,
+    onupdate: Optional[Any] = ...,
+    primary_key: bool = ...,
+    server_default: Optional[_ServerDefaultType[Any]] = ...,
+    server_onupdate: Optional[FetchedValue] = ...,
+    quote: Optional[bool] = ...,
+    unique: Optional[bool] = ...,
+    system: bool = ...,
+    comment: Optional[str] = ...,
+    **kwargs: Any,
+) -> Any: ...
+@overload
+def mapped_column(
+    __name: str,
+    __type: Union[schema._TE, Type[schema._TE]],
+    *args: SchemaEventTarget,
+    autoincrement: Union[bool, Literal["auto", "ignore_fk"]] = ...,
+    default: Optional[Any] = ...,
+    doc: Optional[str] = ...,
+    key: Optional[str] = ...,
+    index: Optional[bool] = ...,
+    info: MutableMapping[Any, Any] = ...,
+    nullable: bool = ...,
+    onupdate: Optional[Any] = ...,
+    primary_key: bool = ...,
+    server_default: Optional[_ServerDefaultType[schema._TE]] = ...,
+    server_onupdate: Optional[FetchedValue] = ...,
+    quote: Optional[bool] = ...,
+    unique: Optional[bool] = ...,
+    system: bool = ...,
+    comment: Optional[str] = ...,
+    **kwargs: Any,
+) -> Any: ...
+@overload
+def mapped_column(
+    __type: Union[schema._TE, Type[schema._TE]],
```

Comment:
Would it be possible to make __type: TypeEngine[X] and then make the function return Mapped[X], so that mapped_column(Integer) -> Mapped[int] automatically, or something similar ?

Reply:
not sure.   my understanding so far has been that pep484 can't propagate a nested type like that.

the things i am thinking about are:

1. Column and all other ORM constructs being treated the same
2. whether to create a new suite of functions altogether
3. starting to introduce mapping straight from annotations

number 3 above is what you are interested in.   the idea is this would be a submarine kind of feature in 1.4 since we are deep into it, then in 2.0 it would be more formally introduced, which is:

```py
@reg.mapped
class MyClass:
    __tablename__ = 'foo'

   id: Mapped[int] = mapped_column(primary_key=True)
   name: Mapped[str]
   fixed_name: Mapped[str] = mapped_column(String(50))
   other_data: Mapped[datetime] = mapped_column(nullable=False)
   binary_data: Mapped[bytes] = mapped_column(mysql.MEDIUMBLOB)
   related: Mapped[List["Related"]] = relationship()

   
```

it should be apparent what we're doing above which is semi-ripping off dataclasses, and adding to declarative the ability to link the annotated types to a pre-fab typeengine.    one challenge is that MySQL doesn't allow VARCHAR without a length so we'd have to introduce some "max length" construct to that dialect and maybe others.

Reply:
OK I think having it return Mapped[_TE] works?   at least the IDE is cool with it lets see what mypy says

Reply:
oh wait, it doesn't.  not exactly.   IMO I thnk pep484 should make this possible but im not sure that it does, there's likely some comp sci name for what this pattern is

Reply:
ok it does.  im not sure why things werent done this way w/ the stubs, but I guess it's because of the mismatch with Column() etc.    OK redefined functions that return someting new might be how to go w/this

all day today ive been looking at how we probably dont need the mypy plugin for most things except the constructor

![Screenshot from 2021-09-10 16-09-09](https://user-images.githubusercontent.com/128223/132911859-5d138842-8c80-4edb-bc8b-bda7d6619b59.png)

Reply:
ill update the issue with the latest thing

Reply:
Did you also consider the possibility of using @overload to automatically handle nullable= -> Optional[] argument to Column, and/or the collection_class= argument so that the type of collection being mapped is inferred automatically?

Those would be nice to have and to avoid repeating the same information twice, once in the column arguments and a second time in the type info.

Otherwise it's fantastic! It would now really be nice if the __dataclass__ transform proposal referenced earlier could support generating the init method and the plugin can then probably go away. The only blocking point there I can think of is the duality between class level and object level descriptor sqlalchemy uses. Maybe this could be added to the proposal somehow..

Reply:
Regarding optional, IIRC the idea was that it should stay in all cases, since even with nullable=False when creating the object in the python side the attributes may be None

Reply:
> Did you also consider the possibility of using @overload to automatically handle nullable= -> Optional[] argument to Column, and/or the collection_class= argument so that the type of collection being mapped is inferred automatically?

so with this *new* system, where we are actually looking at __annotations__, we might have more options here to divine things from the left hand side.  It looks like Mapped returns an instance of _T in all cases, so right off, all of the examples I'm working on so far are not actually complying with the fact that these attributes can be None as @CaselIT  mentions, because I'm not using `Mapped[Optional[str]]` etc. (see https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html#what-the-plugin-does)   .  so it might be reasonable if in fact someone is putting all `Mapped[str]` etc without `Optional`, then yes we can actually say, OK this is nullable=False.      at the same time though, if someone wants more accurate Python behavior where the attribute can be None in the interpreter even though column is not nullable in the DB, that case is still not really handled smoothly, not really sure.   already im a little concerned that my examples all plowed ahead without `Optional` and i even forgot that this isn't actually accurate in-interpreter behavior. 


> 
> Otherwise it's fantastic! It would now really be nice if the **dataclass** transform proposal referenced earlier could support generating the init method and the plugin can then probably go away.

sorry, im forgetting where that is, i was travelling for most of this thread.   is the proposal to trick the typing tools into thinking we're a dataclass so that their harcoded init rules go into effect?   im not sure that even works with `Mapped[]` ?

>  The only blocking point there I can think of is the duality between class level and object level descriptor sqlalchemy uses. Maybe this could be added to the proposal somehow..

yeah, that.   im not convinced that tricking dataclasses is the way to go here yet.   I would think python would someday add some new more generic system to stdlib, i.e. more generic than dataclasses,  that allows this sort of thing, that is, attributes that generate a constructor.

Reply:
so there's "column(nullable=False)" yielding Mapping[] and not Mapping[Optional[]], then there's "Mapping[] vs. Mapping[Optional[]]" setting the nullable flag.   Im trying to see if `@overloads` can do the first version but im not sure.

Reply:
i dont really know how to do it with `@overloads` ,mypy itself doesnt seem to do as well with typing as pylance in any case and i have to keep restarting vscode everytime i change the .pyi file, is there a hotkey in vscode to reload stubs ?

Reply:
pylance seems to work a lot better than mypy is this a thing?

Reply:
latest version has an overloads for nullable

Reply:
> 
> 
> pylance seems to work a lot better than mypy is this a thing?

it does not seem a high bar to pass

Reply:
in my experience pylance works faster and better than mypy, of course it's all subjective :)

Reply:
You can command palette-> Reload window to 'soft-restart' vscode, it'll reload the extensions. Not sure if there's a specific way to just reload stubs.

Reply:
The dataclass transform thing was here: https://github.com/microsoft/pyright/discussions/1782 and https://github.com/microsoft/pyright/blob/main/specs/dataclass_transforms.md

Reply:
Regarding the Optional thing, it's true that just after creating a new object, the fields may be None if they weren't passed a value in the generated init, however in the more general case, when receiving an object from the database, one can assume that if the column is not nullable, there will be some kind of value. Ideally the init would make the non-nullable fields mandatory, but that would be quite a breaking change. Practically speaking it would be annoying if every usage of an object field that is non-nullable, would be squiggly because maybe it's None, this would generate a lot of noise and would lead to turning the thing off I suspect.